clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndMetadata.java

```diff
@@ -62,8 +64,14 @@ public boolean equals(Object o) {
         OffsetAndMetadata that = (OffsetAndMetadata) o;
 
         if (offset != that.offset) return false;
-        return metadata == null ? that.metadata == null : metadata.equals(that.metadata);
 
+        // An empty metadata gets serialized  and stored as an empty string. Depending on where this call is made, we
+        // may be comparing a purely client side object -- where the value is actually null -- with a version fetched
+        // from the server, where the empty value will be an empty string. So we need to account for both.
+        if (metadata == null || metadata.equals(OffsetFetchResponse.NO_METADATA))
```

Comment:
Mentioned offline, we can probably move move the conversion to empty string into the `OffsetAndMetadata` constructor so that we always handle this consistently.